internal/sops/azkv/keysource.go

```diff
@@ -0,0 +1,195 @@
+	if s.ActiveDirectoryEndpoint != "" {
+		return s.ActiveDirectoryEndpoint
+	}
+	return azure.PublicCloud.ActiveDirectoryEndpoint
+}
+
+// EncryptedDataKey returns the encrypted data key this master key holds.
+func (key *MasterKey) EncryptedDataKey() []byte {
+	return []byte(key.EncryptedKey)
+}
+
+// SetEncryptedDataKey sets the encrypted data key for this master key.
+func (key *MasterKey) SetEncryptedDataKey(enc []byte) {
+	key.EncryptedKey = string(enc)
+}
+
+// Encrypt takes a SOPS data key, encrypts it with Key Vault and stores the result in the EncryptedKey field.
+func (key *MasterKey) Encrypt(plaintext []byte) error {
+	c := newKeyvaultClient(key.authorizer())
+
+	data := base64.RawURLEncoding.EncodeToString(plaintext)
+	p := keyvault.KeyOperationsParameters{Value: &data, Algorithm: keyvault.RSAOAEP256}
+	res, err := c.Encrypt(context.Background(), key.VaultURL, key.Name, key.Version, p)
+	if err != nil {
+		return fmt.Errorf("failed to encrypt data: %w", err)
+	}
+
+	key.EncryptedKey = *res.Result
+	if err != nil {
+		return fmt.Errorf("failed to encrypt data: %w", err)
+	}
+	return nil
+}
+
+// EncryptIfNeeded encrypts the provided SOPS' data key and encrypts it if it hasn't been encrypted yet.
+func (key *MasterKey) EncryptIfNeeded(dataKey []byte) error {
+	if key.EncryptedKey == "" {
+		return key.Encrypt(dataKey)
+	}
+	return nil
+}
+
+// Decrypt decrypts the EncryptedKey field with Azure Key Vault and returns the result.
+func (key *MasterKey) Decrypt() ([]byte, error) {
+	c := newKeyvaultClient(key.authorizer())
+
+	result, err := c.Decrypt(context.Background(), key.VaultURL, key.Name, key.Version, keyvault.KeyOperationsParameters{
+		Algorithm: keyvault.RSAOAEP256, Value: &key.EncryptedKey,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to decrypt data: %w", err)
+	}
+
+	plaintext, err := base64.RawURLEncoding.DecodeString(*result.Result)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decrypt data: %w", err)
+	}
+	return plaintext, nil
+}
+
+// NeedsRotation returns whether the data key needs to be rotated or not.
+func (key *MasterKey) NeedsRotation() bool {
+	return key.token.Token().IsExpired()
+}
+
+// ToString converts the key to a string representation.
+func (key *MasterKey) ToString() string {
+	return fmt.Sprintf("%s/keys/%s/%s", key.VaultURL, key.Name, key.Version)
+}
+
+// ToMap converts the MasterKey to a map for serialization purposes.
+func (key MasterKey) ToMap() map[string]interface{} {
+	out := make(map[string]interface{})
+	out["vaultUrl"] = key.VaultURL
+	out["key"] = key.Name
+	out["version"] = key.Version
+	out["enc"] = key.EncryptedKey
+	return out
+}
+
+func (key *MasterKey) authorizer() autorest.Authorizer {
+	if key.token == nil {
+		return &autorest.NullAuthorizer{}
+	}
+	return autorest.NewBearerAuthorizer(key.token)
+}
+
+func newKeyvaultClient(authorizer autorest.Authorizer) keyvault.BaseClient {
+	c := keyvault.New()
+	c.Authorizer = authorizer
```

Comment:
The base client should be configured with proper rate limit back off. It sounds weird but it is totally possible to hit the rate limits for Key Vault, especially in poorly designed systems that share vaults. Adding back off settings alleviates a lot of the pressure. In some situations it becomes an infinite loop of throttling if you have multiple clients that are all just retrying rapidly.

More details can be found in the docs.
https://docs.microsoft.com/en-us/azure/key-vault/general/overview-throttling